Check for and use return value of fread

https://github.com/andy5995/canfigger/blob/bb21c3ec1be2f679e8315847e439f1510445f844/canfigger.c#L293

I got the warning when building the [AUR package](https://aur.archlinux.org/packages/canfigger):

```
./canfigger-0.3.0/canfigger.c:293:3: warning: ignoring return value of ‘fread’ declared with attribute ‘warn_unused_result’ [-Wunused-result]
  293 |   fread(buffer, 1, file_size, fp);
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

I have warning_level set to 3 in meson. I added the flag manually after I saw this message and ran setup and ninja again. It doesn't show up except for when I build the AUR package with `makepkg`. 

Why?

--- a/canfigger.c
+++ b/canfigger.c
@@ -275,12 +275,19 @@ read_entire_file(const char *filename)
   FILE *fp = fopen(filename, "r");
   if (!fp)
   {
-    fprintf(stderr, "Failed to open %s:%s\n", filename, strerror(errno));
+    fprintf(stderr, "Failed to open %s: %s\n", filename, strerror(errno));
     return NULL;
   }
 
   fseek(fp, 0, SEEK_END);
   long file_size = ftell(fp);
+  if (file_size < 0)
+  {
+    fprintf(stderr, "Error getting the size of %s: %s\n", filename,
+            strerror(errno));
+    fclose(fp);
+    return NULL;
+  }
   fseek(fp, 0, SEEK_SET);
 
   char *buffer = malloc_wrap(file_size + 1);
@@ -290,17 +297,30 @@ read_entire_file(const char *filename)
     return NULL;
   }
 
-  fread(buffer, 1, file_size, fp);
-  buffer[file_size] = '\0';
+  size_t n_bytes = fread(buffer, 1, file_size, fp);
 
-  int r = ferror(fp);
-  clearerr(fp);
-  fclose(fp);
+  if (ferror(fp))
+  {
+    fprintf(stderr, "Error reading %s: %s\n", filename, strerror(errno));
+    free(buffer);
+    fclose(fp);
+    return NULL;
+  }
 
-  if (r == 0)
+  // Note that if the return value of ftell() is -1 this cast would be bad.
+  // However, above, the return value of ftell() is checked, and the function
+  // returns if the value is < 0
+  if (n_bytes == (size_t) file_size)
+  {
+    buffer[file_size] = '\0';
+    fclose(fp);
     return buffer;
+  }
 
-  fprintf(stderr, "Error reading %s(%d)\n", filename, r);
+  free(buffer);
+  fprintf(stderr, "Partial read of %s: expected %ld bytes, got %zu bytes\n",
+          filename, file_size, n_bytes);
+  fclose(fp);
   return NULL;
 }
 

